tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java

```diff
@@ -0,0 +1,256 @@
+
+	private byte[] commaByte;
+	private byte[] bracketOpenByte;
+	private byte[] bracketCloseByte;
+
+	/**
+	 * @throws IOException              File not found or couldn't access file
+	 * @throws IllegalArgumentException Log file is not defined in configuration
+	 */
+	public JsonWriter() throws IOException {
+		this(Collections.<String, String>emptyMap());
+	}
+
+	/**
+	 * @param properties Configuration for writer
+	 * @throws IOException              File not found or couldn't access file
+	 * @throws IllegalArgumentException Log file is not defined in configuration
+	 */
+	public JsonWriter(final Map<String, String> properties) throws IOException {
+		tokens = createTokens(properties);
+		charset = getCharset(properties);
+		commaByte = ",".getBytes(charset);
+		bracketOpenByte = "[".getBytes(charset);
+		bracketCloseByte = "]".getBytes(charset);
+
+		String fileName = getFileName(properties);
+		File file = new File(fileName).getAbsoluteFile();
+		file.getParentFile().mkdirs();
+
+		boolean append = Boolean.parseBoolean(properties.get("append"));
+		FileOutputStream stream = new FileOutputStream(file, append);
+
+		fileChannel = stream.getChannel();
+
+		writer = new OutputStreamWriter(stream);
+
+		boolean buffered = Boolean.parseBoolean(properties.get("buffered"));
+		if (buffered) {
+			writer = new BufferedWriterDecorator(writer);
+		}
+
+		boolean writingThread = Boolean.parseBoolean(properties.get("writingthread"));
+		if (!writingThread) {
+			writer = new SynchronizedWriterDecorator(writer, stream);
+		}
+		FileInputStream inputStream = null;
+		try {
+			inputStream = new FileInputStream(file);
+			inputChannel = inputStream.getChannel();
+			preprocessFile(append);
+		} finally {
+			if (inputStream != null) {
+				inputStream.close();
+			}
+		}
+
+	}
+
+	private Charset getCharset(final Map<String, String> properties) {
+		String charsetName = properties.get("charset");
+		try {
+			return charsetName == null ? Charset.defaultCharset() : Charset.forName(charsetName);
+		} catch (IllegalArgumentException ex) {
+			InternalLogger.log(Level.ERROR, "Invalid charset: " + charsetName);
+			return Charset.defaultCharset();
+		}
+	}
+
+	private String getFileName(final Map<String, String> properties) {
+		String fileName = properties.get("file");
+		if (fileName == null) {
+			throw new IllegalArgumentException("File name is missing for file writer");
+		}
+		return fileName;
+	}
+
+	@Override
+	public void write(final LogEntry logEntry) throws IOException {
+		if (builder == null) {
+			builder = new StringBuilder();
+		} else {
+			builder.setLength(0);
+		}
+		builder.append("{");
+		for (int i = 0; i < tokens.size(); i++) {
+			Token token = tokens.get(i);
+			token.render(logEntry, builder);
+			if (i + 1 < tokens.size()) {
+				builder.append(",");
+			}
+		}
+		builder.append("}");
+		builder.append(",");
+
+		escapeCharacter("\\", "\\\\");
+		escapeCharacter(NEW_LINE, "\\n");
+		escapeCharacter("\t", "\\t");
```

Comment:
I think we need also to escape quotation marks: `"`

Reply:
This one is a bit tricky. At this point, all tokens have rendered the properties, including quotation marks like those in `"message":"asdf"`. Replacing all quotation marks would result in escaping the ones in the properties too, like `\\"message\\":\\"\asdf\\"`. Do you have an idea on how to escape only those in the content part of a property?

Reply:
You could store all tokens in a map with field name as key and token as value (exactly as read from the configuration without the JSON extras). Then, you can write the field name with all JSON extras to the string builder while outputting a log entry and only escape characters for the real value as rendered by its token.

Reply:
Great idea! Gonna try it out tomorrow:)

Reply:
I added an implementation for escaping more special characters, though I had to use another StringBuilder. Performance shouldn't be too worse :confused: